### PR TITLE
(GH-56) OutLineView

### PR DESCRIPTION
**Pull request by jpogran**
This PR implements a DocumentSymbolProvider for the LanguageServer. It parses Puppet manifests and produce symbols for classes, types, parameters and variables.

- [x] Implement DocumentSymbolProvider
  - [x] Classes
  - [x] Types
  - [x] Parameters
  - [x] Variables 
- [x] Move tests out from puppet_parser_spec and to provider_spec
- [x] Implement pending tests
- [x] Fixup commits

### CODE REVIEW

**Comment by jpogran** [comment]
@glennsarti Initial commits squashed, tests fixed. First pass attempt at making rubocop happy. Still todo: determine symbol icons and remove debug list of symbols

**Comment by jpogran** [comment]
https://github.com/lingua-pupuli/puppet-vscode/pull/395 has been merged so this PR can be used by the client now.

File: spec/languageserver/unit/puppet-languageserver/document_symbol_provider_spec.rb [hunk 1]
```diff
@@ -22,8 +22,10 @@
   end
 end
 
-describe 'PuppetLanguageServer::PuppetParserHelper' do
-  let(:subject) { PuppetLanguageServer::PuppetParserHelper }
+describe 'PuppetLanguageServer::Manifest::DocumentSymbolProvider',
+  :if => Gem::Version.new(Puppet.version) >= Gem::Version.new('5.0.0') do
```
Comment (glennsarti) [inline spec/languageserver/unit/puppet-languageserver/document_symbol_provider_spec.rb]: What happens on Puppet 4?  Curious why we aren't testing on it?

**Comment by glennsarti** [comment]
@jpogran  I've update your branch with tests for Puppet 4 and some minor fixes.

**Comment by jpogran** [comment]
Squashed commits and updated changelog

**Comment by jpogran** [comment]
@glennsarti @michaeltlombardi I think we're good to merge this in today

File: lib/puppet-languageserver/message_router.rb [hunk 2]
```diff
@@ -169,6 +169,21 @@ def receive_request(request)
           request.reply_result(nil)
         end
 
+      when 'textDocument/documentSymbol'
+        file_uri = request.params['textDocument']['uri']
+        content  = documents.document(file_uri)
+        begin
+          case documents.document_type(file_uri)
+          when :manifest
+            r = PuppetLanguageServer::PuppetParserHelper.extract_document_symbols(content)
```
Comment (glennsarti) [inline lib/puppet-languageserver/message_router.rb]: I'm suprised rubocop didn't complain about the single letter variable.

File: lib/puppet-languageserver/puppet_parser_helper.rb [hunk 3]
```diff
@@ -222,5 +222,9 @@ def self.compile_to_pretty_relationship_graph(manifest, prioritizer = Puppet::Gr
 
       graph
     end
+
+    def self.extract_document_symbols(content)
```
Comment (glennsarti) [inline lib/puppet-languageserver/puppet_parser_helper.rb]: This shouldn't be here

File: lib/puppet-languageserver/message_router.rb (hunk 2 above)
Comment (glennsarti) [inline lib/puppet-languageserver/message_router.rb]: This should be using ` PuppetLanguageServer::Manifest::DocumentSymbolProvider`

**Comment by jpogran** [comment]
@glennsarti please review...again

**Pull request merged**